File: python/src/bench_sdk/otel.py
"""OpenTelemetry bridge: forward spans from any instrumented framework to Bench.

Frameworks such as OpenAI Agents, Strands, LangChain/LangGraph, Pydantic AI,
CrewAI, LlamaIndex, Google ADK and AutoGen already describe what ran through
OpenTelemetry GenAI spans. Attach this exporter and Bench receives the same
agent/model/tool tree, in metadata-only form, without per-framework adapters.
No paid evaluation starts from receiving spans.
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Sequence

from .client import Bench


def _iso(nanoseconds: int | None) -> str:
    if not nanoseconds:
        return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    return datetime.fromtimestamp(nanoseconds / 1e9, timezone.utc).isoformat().replace("+00:00", "Z")


def span_to_bench(span: Any) -> dict:
    """Convert a finished OpenTelemetry ReadableSpan to record_external_span arguments."""
    context = span.get_span_context() if hasattr(span, "get_span_context") else span.context
    parent = getattr(span, "parent", None)
    attributes = dict(getattr(span, "attributes", None) or {})
    kind = attributes.pop("bench.kind", None)
    status = getattr(getattr(span, "status", None), "status_code", None)
    model = attributes.get("gen_ai.request.model")
    return {
        "trace_id": format(context.trace_id, "032x"),
        "span_id": format(context.span_id, "016x"),
        "parent_span_id": format(parent.span_id, "016x") if parent is not None and getattr(parent, "span_id", 0) else None,
        "name": span.name,
        "kind": kind if isinstance(kind, str) else None,
        "started_at": _iso(getattr(span, "start_time", None)),
        "ended_at": _iso(getattr(span, "end_time", None)),
        "status": "error" if getattr(status, "name", "") == "ERROR" else "ok",
        "attributes": attributes,
        "model": model if isinstance(model, str) else None,
    }


def _result(ok: bool) -> Any:
    try:
        from opentelemetry.sdk.trace.export import SpanExportResult
    except ImportError:  # the bridge is also usable from tests without the OpenTelemetry SDK
        return 0 if ok else 1
    return SpanExportResult.SUCCESS if ok else SpanExportResult.FAILURE


class BenchSpanExporter:
    """An opentelemetry.sdk.trace.export.SpanExporter that records through a Bench client."""

    def __init__(self, bench: Bench):
        self._bench = bench

    def export(self, spans: Sequence[Any]) -> Any:
        for span in spans:
            try:
                self._bench.record_external_span(**span_to_bench(span))
            except Exception:
                pass  # the application's tracer must never fail because of Bench
        return _result(True)

    def shutdown(self) -> None:
        self._bench.flush()

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        self._bench.flush()
        return True


def attach(bench: Bench, provider: Any = None) -> Any:
    """Register the exporter on the global (or given) OpenTelemetry tracer provider.

    Requires the opentelemetry-sdk package, which instrumented frameworks already
    bring. Returns the provider. Spans are exported as they end; the Bench client
    batches delivery itself.
    """
    from opentelemetry import trace
    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import SimpleSpanProcessor

    provider = provider or trace.get_tracer_provider()
    if not isinstance(provider, TracerProvider):
        provider = TracerProvider()
        trace.set_tracer_provider(provider)
    provider.add_span_processor(SimpleSpanProcessor(BenchSpanExporter(bench)))
    return provider

File: src/otel.ts
// OpenTelemetry bridge: forward spans from any instrumented framework to Bench.
//
// The Vercel AI SDK (experimental_telemetry), Mastra, OpenAI Agents, LangChain,
// Strands and other frameworks already describe what ran through OpenTelemetry
// GenAI spans. Register this exporter on the application's tracer provider and
// Bench receives the same agent/model/tool tree, metadata-only, without a
// per-framework adapter. Structural types only: no OpenTelemetry dependency.
import type { Bench, ExternalSpanInput } from './index.js';

type HrTime = [number, number];
/** The subset of @opentelemetry/sdk-trace-base ReadableSpan this bridge reads (SDK 1.x and 2.x). */
export interface ReadableSpanLike {
 name: string;
 spanContext(): { traceId: string; spanId: string };
 parentSpanContext?: { spanId: string };
 parentSpanId?: string;
 startTime: HrTime;
 endTime: HrTime;
 status: { code: number; message?: string };
 attributes: Record<string, unknown>;
 events?: { name: string; attributes?: Record<string, unknown> }[];
}
const toMillis = ([seconds, nanos]: HrTime) => seconds * 1000 + nanos / 1e6;

export function spanToBench(span: ReadableSpanLike): ExternalSpanInput {
 const context = span.spanContext();
 const attributes = { ...span.attributes };
 const kind = attributes['bench.kind'];
 delete attributes['bench.kind'];
 const failed = span.status?.code === 2; // SpanStatusCode.ERROR
 if (failed) {
  // Keep what went wrong: the status message and the recorded exception, so
  // Bench can show the error rather than just an error flag.
  const clip = (value: unknown) => (typeof value === 'string' && value.trim() ? value.trim().slice(0, 2000) : undefined);
  const message = clip(span.status?.message);
  if (message && attributes['error.message'] === undefined) attributes['error.message'] = message;
  for (const event of span.events ?? []) {
   if (event.name !== 'exception') continue;
   for (const key of ['exception.type', 'exception.message'] as const) {
    const value = clip(event.attributes?.[key]);
    if (value && attributes[key] === undefined) attributes[key] = value;
   }
  }
 }
 const model = attributes['gen_ai.request.model'];
 return {
  traceId: context.traceId, spanId: context.spanId,
  parentSpanId: span.parentSpanContext?.spanId ?? span.parentSpanId,
  name: span.name,
  kind: typeof kind === 'string' ? (kind as ExternalSpanInput['kind']) : undefined,
  startedAt: toMillis(span.startTime), endedAt: toMillis(span.endTime),
  status: failed ? 'error' : 'ok',
  attributes, model: typeof model === 'string' ? model : undefined,
 };
}

/** Implements the SpanExporter interface of @opentelemetry/sdk-trace-base. */
export class BenchSpanExporter {
 constructor(private readonly bench: Bench) {}
 export(spans: ReadableSpanLike[], resultCallback: (result: { code: number; error?: Error }) => void): void {
  for (const span of spans) {
   try { this.bench.recordExternalSpan(spanToBench(span)) } catch { /* the application's tracer must never fail because of Bench */ }
  }
  resultCallback({ code: 0 }); // ExportResultCode.SUCCESS
 }
 async shutdown(): Promise<void> { await this.bench.flush() }
 async forceFlush(): Promise<void> { await this.bench.flush() }
}
